text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,238 @@
+- This enables third parties such as cloud providers to scan your binaries for you. Google Cloud [already provides such a service](https://cloud.google.com/container-registry/docs/get-image-vulnerabilities), Amazon has [an open-source project you can deploy](https://aws.amazon.com/blogs/publicsector/detect-vulnerabilities-in-the-docker-images-in-your-applications/) while Azure [integrates several partner solutions](https://docs.microsoft.com/en-us/azure/security-center/security-center-vulnerability-assessment-recommendations). They do not support this specific format yet, but integration into Trivy was very easy, so adding support will likely be trivial.
+
+## Why enable this by default?
+
+If you have 10 programs on your computer that use a library exploitable over the network, your computer and all your data are vulnerable as long as even **a single program** remains vulnerable.
+
+It does not matter if you have discovered and patched 9 out of 10. As long as a single unpatched binary is running, the system is vulnerable.
+
+An attacker needs to only find a single vulnerable program; defense is all-or-nothing. **Every single** vulnerable binary needs to be found and fixed.
+
+If the mechanism proposed in this RFC is not enabled by default, it will not let you discover all vulnerable binaries in the system, and will not be effective at preventing vulnerabiltiy exploitation.
+
+## Alternatives
+
+- Do nothing.
+  - Identifying vulnerable binaries will remain impossible. We will see increasing number of known vulnerabilities unpatched in production.
+- Track version information separately from the binaries, recording it when running `cargo install` and surfacing it through some other Cargo subcommand. When installing not though `cargo install`, rely on Linux package managers to track version information.
+  - Identifying vulnerable binaries will remain impossible on all other platforms, as well as on Linux for code compiled with `cargo build`.
+  - Verification by third parties will remain impossible.
+- Record version information in a `&'static str` in the binary instead if ELF sections, with start/stop markers to allow black-box extraction from the outside.
+  - This has been [prototyped](https://github.com/Shnatsel/rust-audit). It has the upside of allowing the binary itself to introspect its version info with little parsing, but the extraction is less efficient, and this is harder to implement and maintain.
+- Record version information in an industry standard SBOM format instead of a custom format.
+  - This has been prototyped, and we've found the existing formats unsuitable. The primary reasons are a significant binary size increase (the existing formats are quite verbose, not designed for this use case) and issues with reproducible builds (they require timestamps).
```

Comment:
Just as an FYI that CycloneDX now has a way to do reproducible builds.
https://github.com/CycloneDX/cyclonedx-property-taxonomy/pull/70
The other argument is still valid.

Reply:
Thanks for the pointer!

I got in touch with some people involved with CycloneDX and I'm investigating if it's possible to select just the right combination of optional fields to have it not blow up the size. 

It is noticeably bigger than the custom format when uncompressed, but gzip does a really good job compacting it. I will revise the RFC with a CycloneDX-compatible format if the prototype shows promise.

Reply:
I'm active in trying to get the https://github.com/CycloneDX/cyclonedx-rust-cargo project going again which is supposed to create CycloneDX SBOMs for Rust projects.
It turns out to be harder than expected so I'm looking at how auditable does it but that doesn't embed all information needed and not all information can be easily recovered from the current JSON format either. I'm torn on how to continue.

But that is a bit separate from this discussion and I don't mean to derail it.